File: chap3/sec3.3/WaveEquation.m
function u=WaveEquation(f,c,tspan,xspan,u0,u0t,gl,gr)
% WAVEEQUATION solves the wave equation with centered finite differences
%   u=WaveEquation(f,c,Tspan,xspan,u0,u0t,gl,gr); solves the wave
%   equation with wave speed c and forcing function f on the domain
%   xspan and initial conditions in the vector u0 and u0t whose length
%   determines the number of spatial gridpoints, and Dirichlet
%   boundary conditions in the vectors gl and gr whose length
%   determines the number of grid points in time. The corner points
%   are taken from the boundary condition, not the initial condition,
%   to avoid inacuracies due to the implementation of the derivative
%   initial condition. The output matrix u contains the solution in
%   time and space.

J=length(u0)-1; N=length(gl)-1;                     % determine grid
dt=(tspan(2)-tspan(1))/N; dx=(xspan(2)-xspan(1))/J; % mesh parameters
x=(xspan(1):dx:xspan(2)); t=(tspan(1):dt:tspan(2))';
u(1,:)=u0;                                          % store initial data
u(2,:)=u0+dt*u0t;                                   % derivative by Taylor
u(1:N+1,1)=gl(1:end); u(1:N+1,J+1)=gr(1:end);       % store boundary data
e=ones(J-1,1);
for n=2:N,
  u(n+1,2:J)=2*u(n,2:J)-u(n-1,2:J)+c^2*dt^2/dx^2*(u(n,3:J+1)...
    -2*u(n,2:J)+u(n,1:J-1))+dt^2*feval(f,x(2:end-1),t(n));
end;

File: chap3/sec3.3/WaveEquation.py
import numpy as np

def waveEquation(f, c, tspan, xspan, u0, u0t, gl, gr):
    """
    Solves the wave equation with centered finite differences.

    Parameters
    ----------
    f : callable
        Forcing function f(x, t).
    c : float
        Wave speed.
    tspan : list or tuple
        Time interval [t_start, t_end].
    xspan : list or tuple
        Spatial interval [x_start, x_end].
    u0 : ndarray
        Initial condition for u at t = t_start.
    u0t : ndarray
        Initial condition for du/dt at t = t_start.
    gl : ndarray
        Dirichlet boundary condition at the left boundary for all time steps.
    gr : ndarray
        Dirichlet boundary condition at the right boundary for all time steps.

    Returns
    -------
    u : ndarray
        Solution matrix with time along rows and space along columns.
    """
    J = len(u0) - 1  # Number of spatial grid points
    N = len(gl) - 1  # Number of time steps
    dt = (tspan[1] - tspan[0]) / N
    dx = (xspan[1] - xspan[0]) / J

    x = np.linspace(xspan[0], xspan[1], J + 1)
    t = np.linspace(tspan[0], tspan[1], N + 1)

    u = np.zeros((N + 1, J + 1))  # Solution matrix

    # Initial conditions
    u[0, :] = u0
    u[1, :] = u0 + dt * u0t  # Derivative by Taylor series expansion

    # Boundary conditions
    u[:, 0] = gl
    u[:, -1] = gr

    # Solve using centered finite differences
    for n in range(1, N):
        u[n + 1, 1:J] = (
            2 * u[n, 1:J] 
            - u[n - 1, 1:J]
            + (c * dt / dx) ** 2 * (u[n, 2:J + 1] - 2 * u[n, 1:J] + u[n, 0:J - 1])
            + dt ** 2 * f(x[1:-1], t[n])
        )

    return u
